airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -307,63 +362,58 @@ private void updateAlbum(MediaFile file, MusicFolder musicFolder, Instant lastSc
             album.setCoverArtPath(parent.getCoverArtPath());
         }
 
-        boolean firstEncounter = !lastScanned.equals(album.getLastScanned());
-        if (firstEncounter) {
+        if (firstEncounter.get()) {
             album.setFolderId(musicFolder.getId());
-            album.setDurationSeconds(0);
-            album.setSongCount(0);
-            Integer n = albumCount.get(artist);
-            albumCount.put(artist, n == null ? 1 : n + 1);
-        }
-        if (file.getDurationSeconds() != null) {
-            album.setDurationSeconds(album.getDurationSeconds() + file.getDurationSeconds());
-        }
-        if (file.isAudio()) {
-            album.setSongCount(album.getSongCount() + 1);
-        }
-        album.setLastScanned(lastScanned);
-        album.setPresent(true);
-        albumDao.createOrUpdateAlbum(album);
-        if (firstEncounter) {
+            albumCount.computeIfAbsent(artist, k -> new AtomicInteger(0)).incrementAndGet();
             indexManager.index(album);
         }
-
+        
         // Update the file's album artist, if necessary.
         if (!ObjectUtils.equals(album.getArtist(), file.getAlbumArtist())) {
             file.setAlbumArtist(album.getArtist());
             mediaFileDao.createOrUpdateMediaFile(file);
         }
     }
 
-    private void updateArtist(MediaFile file, MusicFolder musicFolder, Instant lastScanned, Map<String, Integer> albumCount) {
+    private void updateArtist(MediaFile file, MusicFolder musicFolder, Instant lastScanned, Map<String, AtomicInteger> albumCount, Map<String, Artist> artists) {
         if (file.getAlbumArtist() == null || !file.isAudio()) {
             return;
         }
 
-        Artist artist = artistDao.getArtist(file.getAlbumArtist());
-        if (artist == null) {
-            artist = new Artist();
-            artist.setName(file.getAlbumArtist());
-        }
+        final AtomicBoolean firstEncounter = new AtomicBoolean(false);
+        
+        Artist artist = artists.compute(file.getAlbumArtist(), (k,v) -> {
+            Artist a = v;
+            
+            if (a == null) {
+                a = artistDao.getArtist(k);
+            }
+            
+            if (a == null) {
+                a = new Artist();
+                a.setName(k);
+            }
+            
+            Integer n = Math.max(Optional.ofNullable(albumCount.get(a.getName())).map(x -> x.get()).orElse(0), Optional.ofNullable(a.getAlbumCount()).orElse(0));
```

Comment:
this is slightly thread-unsafe, but the important thing to note here is that albumCount should be monotonically increasing.

thus if another thread updates it to a higher number than what this thread is going to update it to (maybe this thread was preempted), we need to retain the higher number